### PR TITLE
Modify Synthetic Reward/Behavior Policy Functions

### PR DESCRIPTION
**Pull request by usaito**
## New Features
- define some new functions to generate expected rewards and behavior policy in `obp.dataset.SyntheticBanditDataset`. These new functions provide more natural and flexible reward functions and behavior policies. Specifically, previous reward and behavior policy functions were easily specified by a linear model, which is often unrealistic. Some of the new functions are based on polynomial feature transformations to be more complex, and thus realistic. The added (or modified) functions are as follows.
    - logistic_reward_function
    - logistic_polynomial_reward_function
    - linear_reward_function
    - polynomial_reward_function
    - linear_behavior_policy
    - polynomial_behavior_policy

Note that the new reward functions are based on [`_base_reward_function`](https://github.com/st-tech/zr-obp/blob/feature/synthetic-dataset/obp/dataset/synthetic.py#L663) and behavior policy functions are based on [`_base_behavior_policy_function`](https://github.com/st-tech/zr-obp/blob/418e591fdf16a2cfcdd4ab4f30d92835d1846001/obp/dataset/synthetic.py#L663). These base functions include more detailed description on the inner workings of the above new functions.

- users are now able to specify their own `action_context` as in the following example.
```python
dataset = SyntheticBanditDataset(
    n_actions=10,
    dim_context=10,
    action_context=np.random.normal((10, 5)) # (n_actions, dim_action_context)
    ....
)
```
Note that when `action_context=None`, one-hot action representation will be used as default.

- users are now able to control the optimality and entropy of the behavior policy in a flexible manner.
```python
dataset = SyntheticBanditDataset(
    n_actions=10,
    dim_context=10,
    reward_function=logistic_reward_function,
    beta=1.0, 
    behavior_policy_function=None,
    ....
)
```
When `behavior_policy_function=None` as above, the behavior policy will be generated as follows.
```
\pi(a|x) = softmax (\beta q(x,a))
```
where `beta` is the inverse temperature parameter, which controls the optimality and entropy of the behavior policy.
A large value leads to a near-deterministic behavior policy, while a small value leads to a near-uniform behavior policy.
A positive value leads to a near-optimal behavior policy, while a negative value leads to a sub-optimal behavior policy.

## Minor Fixes
- add some tests to `tests/dataset/test_synthetic.py` regarding the new features
- add action_dist of pi_b to `bandit_feedback` (output of the `obtain_batch_bandit_feedback` method)
- avoid the overflow warning in the logit function using `scipy.special.logit`
- fix some tests regarding OPE performance. In particular, define [`NaiveEstimator`](https://github.com/st-tech/zr-obp/blob/418e591fdf16a2cfcdd4ab4f30d92835d1846001/tests/ope/test_offline_estimation_performance.py#L101) instead of RandomOffPolicyEstimator, which has a randomness in the estimation and thus is weird as a baseline estimator.
- fix some typos that failed to be addressed in the previous PR (https://github.com/st-tech/zr-obp/pull/144)

### CODE REVIEW

File: obp/dataset/synthetic.py [hunk 1]
```diff
@@ -389,71 +442,294 @@ def linear_reward_function(
...
+    """
+    check_scalar(degree, "degree", int, min_val=1)
     check_array(array=context, name="context", expected_dim=2)
     check_array(array=action_context, name="action_context", expected_dim=2)
 
+    poly = PolynomialFeatures(degree=degree)
+    context_ = poly.fit_transform(context)
+    action_context_ = poly.fit_transform(action_context)
+    datasize, context_dim = context_.shape
+    n_actions, action_context_dim = action_context_.shape
```
[the remark this answers is not in the document]
Comment (aiueola) [reply obp/dataset/synthetic.py:574]: The docstrings also use `(n_rounds, dim_context)` and `(n_actions, dim_action)`.

File: obp/dataset/synthetic.py [hunk 2]
```diff
@@ -389,71 +442,294 @@ def linear_reward_function(
...
+        random_state=random_state,
+    )
+
+
+def _base_reward_function(
+    context: np.ndarray,
+    action_context: np.ndarray,
+    degree: int = 3,
+    random_state: Optional[int] = None,
+) -> np.ndarray:
```
[the remark this answers is not in the document]
Comment (usaito) [reply obp/dataset/synthetic.py:516]: @aiueola Agree. `effective_dim_context` seems more appropriate wording here.

[the remark this answers is not in the document]
Comment (usaito) [reply obp/dataset/synthetic.py:516]: But, in my understanding, this is not for computational tractability, as we still need to generate `context_`, which could be huge. Rather, I think your suggestion is useful to create another type of misspecification, which makes the reward modeling more realistic, as in the real-world, there should always be some irrelevant feature dimensions.
